[BOLT][heatmap] Produce zoomed-out heatmaps

aaupov opened the pull request:
Add a capability to produce multiple heatmaps with given bucket sizes.

The default heatmap block size (64B) could be too fine-grained for
large binaries. Extend the option `block-size` to accept a list of
bucket sizes for additional heatmaps with coarser granularity. The
heatmap is simply rescaled so provided sizes should be multiples of
each other. Human-readable suffixes can be used, e.g. 4K, 16kb, 1MiB.

New defaults: 64B (base bucket size), 4KB (default page size),
256KB (for large binaries).

Test Plan: updated heatmap-preagg.test

aaupov committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.4

aaupov committed:
drop alignTo

Created using spr 1.3.4

aaupov committed:
fix test

Created using spr 1.3.4

paschalis-mpeis reviewed (commented):
Great option! It may be worth listing this in the [Heatmaps.md](https://github.com/llvm/llvm-project/blob/main/bolt/docs/Heatmaps.md), alongside the suggested usage.

> Suggested values to use: 4096 (default page size), 16384 (16k page),
1048576 (1MB for XL workloads).

aaupov committed:
add new option to Heatmaps.md

Created using spr 1.3.4

aaupov committed:
suggested values

Created using spr 1.3.4

maksfb commented:
Instead of introducing the new option, can we reuse `--block-size=` and make it accept multiple values or a different format?

> If rescaling is impossible, no heatmap is produced.

Without a printed warning? I understand that the size limitation is driven by implementation, but at the very least we should report that there is no expected output.

Alternatively, use a different option format that encapsulates the limitations. E.g., `--block-size=<initial_size>{:<scale>:<count>}` or `--block-size=<initial_size>{:<scale1>,<scale2>...}`.

aaupov committed:
Modify block-size option

Created using spr 1.3.4

aaupov committed:
clang-format

Created using spr 1.3.4

maksfb commented on bolt/docs/Heatmaps.md:
````diff
@@ -89,7 +89,12 @@ For the generation, the default bucket size was used with a line size of 128.
 ```
--line-size=<uint>   - number of entries per line (default 256)
+-line-size=<uint>   - number of entries per line (default 256).
+                      Use 128 if the heatmap doesn't fit screen horizontally.
````
"Use a smaller value if ..."?

maksfb reviewed (commented):
The implementation looks good to me.

A couple comments regarding the interface:
  * If we are producing multiple files by default, we might have to change docs in more places.
  * I find the power-of-two limitation for the scale unnecessary.
  * The variance in default scale steps looks a bit unnatural (64, 4, 64 again). I would keep just 64 and 64 making the default sizes 64 bytes, 4 KB, 256 KB.

aaupov commented:
@maksfb:

> * If we are producing multiple files by default, we might have to change docs in more places.

Makes sense. I'll update heatmap doc, anything else?

> * I find the power-of-two limitation for the scale unnecessary.

This made it impossible to enter invalid scales. While working on a custom parser, I realized that a better UX would be to allow specifying bucket sizes in natural units (e.g. 4KB, 1MB):
```
--block-size=default_size[,size1,...]
```

If you agree, I'll switch to that and print warnings in case of invalid sizes/rescales.

> * The variance in default scale steps looks a bit unnatural (64, 4, 64 again). I would keep just 64 and 64 making the default sizes 64 bytes, 4 KB, 256 KB.

It's the downstream of bucket size selection. 1MB works really well for XL binaries, to have a high-level view of the whole address space. I'd like to keep 1MB for that reason. We can add 256KB for sure. 

LMK if that sounds good.

maksfb commented:
> @maksfb:
> 
> > * If we are producing multiple files by default, we might have to change docs in more places.
> 
> Makes sense. I'll update heatmap doc, anything else?

`llvm-bolt-heatmap` usage message if possible.

> > * I find the power-of-two limitation for the scale unnecessary.
> 
> This made it impossible to enter invalid scales. While working on a custom parser, I realized that a better UX would be to allow specifying bucket sizes in natural units (e.g. 4KB, 1MB):
> 
> ```
> --block-size=default_size[,size1,...]
> ```
> 
> If you agree, I'll switch to that and print warnings in case of invalid sizes/rescales.

Okay. Then the burden falls on the proper UI messaging and error/warning diagnostics.

> > * The variance in default scale steps looks a bit unnatural (64, 4, 64 again). I would keep just 64 and 64 making the default sizes 64 bytes, 4 KB, 256 KB.
> 
> It's the downstream of bucket size selection. 1MB works really well for XL binaries, to have a high-level view of the whole address space. I'd like to keep 1MB for that reason. We can add 256KB for sure.
> 
> LMK if that sounds good.

If 256KB is difficult to read for large binaries, 1MB works for me.

aaupov committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.4

[skip ci]

aaupov committed:
update block-size option

Created using spr 1.3.4

aaupov committed:
update block-size option

Created using spr 1.3.4

aaupov committed:
usage message

Created using spr 1.3.4

aaupov commented:
> > @maksfb:
> > > * If we are producing multiple files by default, we might have to change docs in more places.
> > 
> > 
> > Makes sense. I'll update heatmap doc, anything else?
> 
> `llvm-bolt-heatmap` usage message if possible.

Added usage message:
```
OVERVIEW:  BOLT Code Heatmap tool

  Produces code heatmaps using sampled profile

  Inputs:
  - Binary (supports BOLT-optimized binaries),
  - Sampled profile collected from the binary:
    - perf data or pre-aggregated profile data (instrumentation profile not supported)
    - perf data can have basic (IP) or branch-stack (LBR) samples

  Outputs:
  - Heatmaps: colored ASCII (requires a color-capable terminal or a conversion tool like `aha`)
    Multiple heatmaps are produced by default with different granularities (set by `block-size` option)
  - Section hotness: per-section samples% and utilization%
  - Cumulative distribution: working set size corresponding to a given percentile of samples
```

> 
> > > * I find the power-of-two limitation for the scale unnecessary.
> > 
> > 
> > This made it impossible to enter invalid scales. While working on a custom parser, I realized that a better UX would be to allow specifying bucket sizes in natural units (e.g. 4KB, 1MB):
> > ```
> > --block-size=default_size[,size1,...]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If you agree, I'll switch to that and print warnings in case of invalid sizes/rescales.
> 
> Okay. Then the burden falls on the proper UI messaging and error/warning diagnostics.

Added check for sorted values. Not checking the provided values though.

> 
> > > * The variance in default scale steps looks a bit unnatural (64, 4, 64 again). I would keep just 64 and 64 making the default sizes 64 bytes, 4 KB, 256 KB.
> > 
> > 
> > It's the downstream of bucket size selection. 1MB works really well for XL binaries, to have a high-level view of the whole address space. I'd like to keep 1MB for that reason. We can add 256KB for sure.
> > LMK if that sounds good.
> 
> If 256KB is difficult to read for large binaries, 1MB works for me.

256KB works well for large binaries.

aaupov committed:
enforce block sizes to be multiples

Created using spr 1.3.4

maksfb reviewed (approved):
Thanks for addressing comments.
